File: sdks/typescript/src/__tests__/channels.test.ts
import { describe, it, expect, vi } from 'vitest';
import { ChannelsResource } from '../resources/channels';
import type { HttpClient } from '../utils/http';

// Minimal HttpClient stub: each verb resolves to a preset envelope and records
// the call so we can assert on path/body.
function stubHttp(responses: Record<string, unknown>) {
  const calls: Array<{ verb: string; path: string; body?: unknown }> = [];
  const make = (verb: string) =>
    vi.fn(async (path: string, body?: unknown) => {
      calls.push({ verb, path, body });
      return responses[`${verb} ${path}`] ?? responses[path];
    });
  const http = {
    get: make('GET'),
    post: make('POST'),
    put: make('PUT'),
    patch: make('PATCH'),
    delete: make('DELETE'),
  } as unknown as HttpClient;
  return { http, calls };
}

describe('ChannelsResource', () => {
  it('connect unwraps the envelope and surfaces the QR payload', async () => {
    const { http } = stubHttp({
      'POST /channels/ch1/connect': {
        success: true,
        data: {
          channel: { id: 'ch1', name: 'wa', type: 'whatsapp', connection_status: 'connecting' },
          qr_code: 'QR-PAYLOAD-123',
          expires_in: 60,
        },
      },
    });
    const res = await new ChannelsResource(http).connect('ch1');
    expect(res.qr_code).toBe('QR-PAYLOAD-123');
    expect(res.expires_in).toBe(60);
    expect(res.channel.id).toBe('ch1');
    expect(res.channel.connection_status).toBe('connecting');
  });

  it('create sends credentials and returns the channel', async () => {
    const { http, calls } = stubHttp({
      'POST /channels': {
        success: true,
        data: { id: 'ch9', name: 'wa', type: 'whatsapp', connection_status: 'disconnected' },
      },
    });
    const ch = await new ChannelsResource(http).create({
      name: 'wa',
      type: 'whatsapp',
      credentials: { access_token: 'secret' },
    });
    expect(ch.id).toBe('ch9');
    const body = calls[0].body as { credentials?: Record<string, string> };
    expect(body.credentials?.access_token).toBe('secret');
  });

  it('update uses PUT', async () => {
    const { http, calls } = stubHttp({
      'PUT /channels/ch1': { success: true, data: { id: 'ch1', name: 'renamed' } },
    });
    await new ChannelsResource(http).update('ch1', { name: 'renamed' });
    expect(calls[0].verb).toBe('PUT');
    expect(calls[0].path).toBe('/channels/ch1');
  });

  it('requestPairCode posts phone_number and returns the pair code', async () => {
    const { http, calls } = stubHttp({
      'POST /channels/ch1/pair': { success: true, data: { pair_code: 'ABCD-1234' } },
    });
    const res = await new ChannelsResource(http).requestPairCode('ch1', '+5511999999999');
    expect(res.pair_code).toBe('ABCD-1234');
    expect((calls[0].body as { phone_number: string }).phone_number).toBe('+5511999999999');
  });

  it('list unwraps the plain data array', async () => {
    const { http } = stubHttp({
      'GET /channels': { success: true, data: [{ id: 'a' }, { id: 'b' }] },
    });
    const list = await new ChannelsResource(http).list();
    expect(list.map((c) => c.id)).toEqual(['a', 'b']);
  });
});

File: sdks/go/channels_test.go
package linktor

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/linktor/linktor-go/types"
)

// TestChannelsConnectReturnsQR proves Connect unwraps the {success,data} envelope
// and surfaces the QR payload — the gap that previously made connect() drop the
// QR by deserializing into a bare Channel.
func TestChannelsConnectReturnsQR(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/channels/ch1/connect" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if got := r.Header.Get("X-API-Key"); got != "lk_test" {
			t.Errorf("expected API key header, got %q", got)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, `{"success":true,"data":{
			"channel":{"id":"ch1","name":"wa","type":"whatsapp","connection_status":"connecting"},
			"qr_code":"QR-PAYLOAD-123","expires_in":60}}`)
	}))
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL), WithAPIKey("lk_test"))
	res, err := c.Channels.Connect(context.Background(), "ch1")
	if err != nil {
		t.Fatalf("Connect: %v", err)
	}
	if res.QRCode != "QR-PAYLOAD-123" {
		t.Errorf("QRCode = %q, want QR-PAYLOAD-123", res.QRCode)
	}
	if res.ExpiresIn != 60 {
		t.Errorf("ExpiresIn = %d, want 60", res.ExpiresIn)
	}
	if res.Channel == nil || res.Channel.ID != "ch1" {
		t.Errorf("Channel not populated: %+v", res.Channel)
	}
	if res.Channel != nil && res.Channel.ConnectionStatus != types.ConnectionStatusConnecting {
		t.Errorf("ConnectionStatus = %q, want connecting", res.Channel.ConnectionStatus)
	}
}

// TestChannelsCreateSendsCredentials proves Create serializes the credentials
// map (previously absent from CreateChannelInput) and unwraps the created channel.
func TestChannelsCreateSendsCredentials(t *testing.T) {
	var body map[string]json.RawMessage
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_ = json.NewDecoder(r.Body).Decode(&body)
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, `{"success":true,"data":{"id":"ch9","name":"wa","type":"whatsapp","status":"inactive"}}`)
	}))
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL), WithAPIKey("lk_test"))
	ch, err := c.Channels.Create(context.Background(), &types.CreateChannelInput{
		Name:        "wa",
		Type:        types.ChannelTypeWhatsApp,
		Credentials: map[string]string{"access_token": "secret"},
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if ch.ID != "ch9" {
		t.Errorf("ID = %q, want ch9", ch.ID)
	}
	if _, ok := body["credentials"]; !ok {
		t.Errorf("credentials not sent in request body: %v", body)
	}
}

// TestChannelsRequestPairCode proves pairing posts the phone number and returns
// the pair code.
func TestChannelsRequestPairCode(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/channels/ch1/pair" {
			t.Errorf("unexpected path: %s", r.URL.Path)
		}
		var in types.PairCodeInput
		_ = json.NewDecoder(r.Body).Decode(&in)
		if in.PhoneNumber != "+5511999999999" {
			t.Errorf("phone = %q", in.PhoneNumber)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = io.WriteString(w, `{"success":true,"data":{"pair_code":"ABCD-1234"}}`)
	}))
	defer srv.Close()

	c := NewClient(WithBaseURL(srv.URL), WithAPIKey("lk_test"))
	res, err := c.Channels.RequestPairCode(context.Background(), "ch1", "+5511999999999")
	if err != nil {
		t.Fatalf("RequestPairCode: %v", err)
	}
	if res.PairCode != "ABCD-1234" {
		t.Errorf("PairCode = %q, want ABCD-1234", res.PairCode)
	}
}
